File: java/testsuite/Jogl4Examples/ShadersExample/src/render/OpenGlOfflineSphereRenderer.java
package render;

import java.nio.ByteBuffer;

import com.jogamp.opengl.GL;
import com.jogamp.opengl.GL4;
import com.jogamp.opengl.GLAutoDrawable;
import com.jogamp.opengl.GLCapabilities;
import com.jogamp.opengl.GLDrawableFactory;
import com.jogamp.opengl.GLEventListener;
import com.jogamp.opengl.GLOffscreenAutoDrawable;
import com.jogamp.opengl.GLProfile;

import model.ShadersModel;
import vsdk.toolkit.common.linealAlgebra.Matrix4x4d;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.render.jogl.Jogl4SphereRenderer;

public class OpenGlOfflineSphereRenderer
{
    public RGBImageUncompressed render(
        ShadersModel model,
        Matrix4x4d modelRotation,
        int width,
        int height)
    {
        GLProfile profile = GLProfile.get(GLProfile.GL4);
        GLCapabilities caps = new GLCapabilities(profile);
        caps.setDoubleBuffered(false);
        GLOffscreenAutoDrawable pbuffer = GLDrawableFactory
            .getFactory(profile)
            .createOffscreenAutoDrawable(null, caps, null, width, height);

        OfflineRendererListener listener = new OfflineRendererListener(
            model,
            modelRotation,
            width,
            height);
        pbuffer.addGLEventListener(listener);
        pbuffer.display();
        pbuffer.destroy();
        return listener.getCapturedImage();
    }

    private static final class OfflineRendererListener implements GLEventListener
    {
        private final ShadersModel model;
        private final Matrix4x4d modelRotation;
        private final int width;
        private final int height;
        private RGBImageUncompressed capturedImage;

        private OfflineRendererListener(
            ShadersModel model,
            Matrix4x4d modelRotation,
            int width,
            int height)
        {
            this.model = model;
            this.modelRotation = modelRotation;
            this.width = width;
            this.height = height;
        }

        @Override
        public void init(GLAutoDrawable drawable)
        {
        }

        @Override
        public void display(GLAutoDrawable drawable)
        {
            GL4 gl = drawable.getGL().getGL4();
            gl.glViewport(0, 0, width, height);
            gl.glEnable(GL4.GL_DEPTH_TEST);
            gl.glClearColor(0, 0, 0, 1);
            gl.glClear(GL4.GL_COLOR_BUFFER_BIT | GL4.GL_DEPTH_BUFFER_BIT);

            Jogl4SphereRenderer.draw(
                gl,
                model.getSphere(),
                model.getCamera(),
                model.getLight(),
                model.getActiveMaterialForCurrentShading(),
                model.getQuality(),
                model.getTextureMap(),
                model.getBumpMapHeightRgb(),
                modelRotation,
                model.getSphereMeridians(),
                model.getSphereParallels());

            gl.glFinish();
            capturedImage = captureRgbImage(gl, width, height);
        }

        @Override
        public void dispose(GLAutoDrawable drawable)
        {
            GL4 gl = drawable.getGL().getGL4();
            Jogl4SphereRenderer.dispose(gl);
        }

        @Override
        public void reshape(GLAutoDrawable drawable, int x, int y, int width, int height)
        {
        }

        private static RGBImageUncompressed captureRgbImage(GL4 gl, int width, int height)
        {
            ByteBuffer bb = ByteBuffer.allocateDirect(3 * width * height);
            gl.glPixelStorei(GL.GL_PACK_ALIGNMENT, 1);
            gl.glReadPixels(0, 0, width, height, GL.GL_RGB, GL.GL_UNSIGNED_BYTE, bb);

            RGBImageUncompressed image = new RGBImageUncompressed();
            image.init(width, height);
            int pos = 0;
            for ( int y = image.getYSize() - 1; y >= 0; y-- ) {
                for ( int x = 0; x < image.getXSize(); x++ ) {
                    image.putPixel(x, y, bb.get(pos), bb.get(pos + 1), bb.get(pos + 2));
                    pos += 3;
                }
            }
            return image;
        }

        private RGBImageUncompressed getCapturedImage()
        {
            return capturedImage;
        }
    }
}

File: cpp/testsuite/OpenGL4Examples/ShadersExample/src/render/OpenGlOfflineSphereRenderer.cpp
#include "OpenGlOfflineSphereRenderer.h"
#include "../model/ShadersModel.h"

#include <vector>

#include <GL/glew.h>
#define GLFW_INCLUDE_NONE
#include <GLFW/glfw3.h>

#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/render/opengl4/OpenGL4SphereRenderer.h"
#include "vsdk/toolkit/render/opengl4/OpenGL4ImageRenderer.h"
#include "vsdk/toolkit/render/opengl4/OpenGL4MatrixRenderer.h"

using vsdk::toolkit::render::opengl4::OpenGL4SphereRenderer;
using vsdk::toolkit::render::opengl4::OpenGL4MatrixRenderer;

RGBImageUncompressed* OpenGlOfflineSphereRenderer::render(
    ShadersModel* model,
    const Matrix4x4d& modelRotation,
    int width,
    int height)
{
    if (!model) return 0;
    if (!glfwInit()) return 0;
    glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 1);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
#ifdef __APPLE__
    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GLFW_TRUE);
#endif
    GLFWwindow* window = glfwCreateWindow(width, height, "offline", 0, 0);
    if (!window) { glfwTerminate(); return 0; }
    glfwMakeContextCurrent(window);
    glewExperimental = GL_TRUE;
    if (glewInit() != GLEW_OK) { glfwDestroyWindow(window); glfwTerminate(); return 0; }

    model->camera->updateViewportResize(width, height);
    glViewport(0, 0, width, height);
    glEnable(GL_DEPTH_TEST);
    glClearColor(0, 0, 0, 1);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    OpenGL4SphereRenderer::draw(
        model->sphere,
        model->camera,
        model->light,
        &model->material,
        &model->quality,
        model->textureMap,
        model->bumpMapHeightRgb,
        modelRotation,
        model->sphereMeridians,
        model->sphereParallels);

    glFinish();

    std::vector<unsigned char> bytes((size_t)width * (size_t)height * 3u, 0u);
    glPixelStorei(GL_PACK_ALIGNMENT, 1);
    glReadPixels(0, 0, width, height, GL_RGB, GL_UNSIGNED_BYTE, bytes.data());

    RGBImageUncompressed* image = new RGBImageUncompressed();
    image->init(width, height);
    size_t pos = 0;
    for (int y = height - 1; y >= 0; y--) {
        for (int x = 0; x < width; x++) {
            image->putPixel(x, y, (char)bytes[pos], (char)bytes[pos + 1], (char)bytes[pos + 2]);
            pos += 3;
        }
    }

    OpenGL4SphereRenderer::dispose();
    OpenGL4ImageRenderer::dispose();
    OpenGL4MatrixRenderer::release();
    glfwDestroyWindow(window);
    glfwTerminate();
    return image;
}
